src/librbd/cache/pwl/rwl/WriteLog.cc

```diff
@@ -439,6 +439,19 @@ void WriteLog<I>::load_existing_entries(DeferredContexts &later) {
   this->update_sync_points(missing_sync_points, sync_point_entries, later, MIN_WRITE_ALLOC_SIZE);
 }
 
+template <typename I>
+void WriteLog<I>::calc_allocated_and_cached_data(std::shared_ptr<pwl::GenericLogEntry> log_entry, uint32_t alloc_size) {
+  auto gen_write_entry = static_pointer_cast<GenericWriteLogEntry>(log_entry);
+  if (log_entry->is_write_entry()) {
+    uint64_t bytes_allocated = alloc_size;
+    if (gen_write_entry->ram_entry.write_bytes > bytes_allocated) {
+      bytes_allocated = gen_write_entry->ram_entry.write_bytes;
+    }
+    this->m_bytes_allocated += bytes_allocated;
+    this->m_bytes_cached += log_entry->write_bytes();
```

Comment:
I'm not sure I follow.  The problem is writesame, correct?  For `m_bytes_allocated`, why is `ram_entry.write_bytes` used instead of `write_bytes()`?  In the "writesame these 512 bytes 1000 times" example, should `m_bytes_allocated` be incremented by 512 or 512000?

Reply:
@idryomov IMO it should be `ram_entry.write_bytes` since we don't write all the writesame bytes to the disk. In the example you gave, in `ssd` cache, only 512 bytes is written to the disk, not 512000. But `log_entry->write_bytes()` returns 512000 for writesame[0]

[0]https://github.com/ceph/ceph/blob/master/src/librbd/cache/pwl/LogEntry.h#L204

Reply:
see : https://github.com/ceph/ceph/blob/master/src/librbd/cache/pwl/LogEntry.h#L183-L188

Reply:
@MahatiC My LGTM comment was based on the assumption that in the above example `ram_entry.ws_datalen` would be 512 and `ram_entry.write_bytes` would be 512000.  Then `log_entry->write_bytes()` would return 512.  Am I wrong here?

Reply:
@idryomov hmm, [0] points that `ram_entry.ws_datalen` would be 100 in the example and `ram_entry_write_bytes` [1] would be 512.

[0] https://github.com/ceph/ceph/blob/master/src/librbd/cache/pwl/LogEntry.h#L195
[1]https://github.com/ceph/ceph/blob/master/src/librbd/cache/pwl/LogEntry.h#L31

Reply:
Are you saying that `ram_entry.ws_datalen` is not a byte count but rather a multiplier?  I don't think so -- check out e.g. `rbd_writesame()` signature in `librbd.h`:
```
ssize_t rbd_writesame(rbd_image_t image, uint64_t ofs, size_t len,
                      const char *buf, size_t data_len,
                      int op_flags);
```
`len` is the size of the intended write in bytes, `data_len` is the size of `buf` in bytes.  The implementation in `librbd.cc` then checks that `len % data_len == 0`.  The multiplier (i.e. `len / data_len`) is never stored explicitly.

Further, `len` may get clipped in `clip_io()` so `len / data_len` isn't guaranteed to be a whole number...

Reply:
@MahatiC ping

Reply:
@idryomov . @MahatiC last week quit intel.

Reply:
Ah, she probably has other things on her mind right now then ;)  I'll consider this conversation resolved as I think we all agree that `ram_entry.ws_datalen` is always a byte count and never a multiplier.